### PR TITLE
feat: Make cluster cache sync more robust

### PR DESCRIPTION
**Pull request by jannfis**
This is a follow-up to https://github.com/argoproj/argo-cd/pull/8404 and makes use of https://github.com/argoproj/gitops-engine/pull/374 

It exposes the possibility to make the cluster cache sync process more reliable on flaky or otherwise unreliable network connections, as can be found on the path to systems at the far edge.

Retry can be configured using following environment variables: 

* `ARGOCD_CLUSTER_CACHE_ATTEMPT_LIMIT` - (number) Number of maximum attempts for resource list requests (defaults to `1`)
* `ARGOCD_CLUSTER_CACHE_RETRY_USE_BACKOFF` - (boolean) Whether to use a backoff strategy for retries (defaults to `false`)

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

File: controller/cache/cache.go [hunk 1]
```diff
@@ -62,13 +71,22 @@ var (
...
+	// clusterCacheRetryUseBackoff specifies whether to use a backoff strategy on cluster cache sync, if retry is enabled
+	clusterCacheRetryUseBackoff bool = false
 )
 
 func init() {
 	clusterCacheResyncDuration = env.ParseDurationFromEnv(EnvClusterCacheResyncDuration, clusterCacheResyncDuration, 0, math.MaxInt64)
 	clusterCacheWatchResyncDuration = env.ParseDurationFromEnv(EnvClusterCacheWatchResyncDuration, clusterCacheWatchResyncDuration, 0, math.MaxInt64)
 	clusterCacheListPageSize = env.ParseInt64FromEnv(EnvClusterCacheListPageSize, clusterCacheListPageSize, 0, math.MaxInt64)
 	clusterCacheListSemaphoreSize = env.ParseInt64FromEnv(EnvClusterCacheListSemaphore, clusterCacheListSemaphoreSize, 0, math.MaxInt64)
+	clusterCacheRetryLimit = int32(env.ParseInt64FromEnv(EnvClusterCacheRetryLimit, 1, 1, math.MaxInt32))
```
Comment (jannfis) [inline controller/cache/cache.go:88]: The upper limit of `ParseInt64FromEnv()` actually makes the conversion to int32 safe.

File: controller/cache/cache.go [hunk 2]
```diff
@@ -62,13 +71,22 @@ var (
 	// clusterCacheListPageSize is the page size when performing K8s list requests.
 	// 500 is equal to kubectl's size
 	clusterCacheListPageSize int64 = 500
+
+	// clusterCacheRetryLimit sets a retry limit for failed requests during cluster cache sync
+	// If set to 1, retries are disabled.
+	clusterCacheRetryLimit int32 = 1
```
Comment (crenshaw-dev) [inline controller/cache/cache.go:77]: The tiniest of nitpicks: I'd expect "retry limit" to mean "limit of the number of retries" where the 2nd request is the first retry. In other words, I'd expect retryLimit=0 to mean "no retries."

I think if we wanted 1 to mean "no retries," we should call the env var "attempt limit."

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
I'm sure you thought about this, but is it possible to mock enough stuff to test a retry situation?

File: controller/cache/cache.go (hunk 2 above)
Comment (jannfis) [reply controller/cache/cache.go:77]: That's actually a good suggestion, thanks :+1: Will change the naming. Although the appropriate changes have been named in GitOps engine like this already :\

**Comment by jannfis** [comment]
> I'm sure you thought about this, but is it possible to mock enough stuff to test a retry situation?

The mock should imho live in GitOps Engine, and I agree, it probably should be mocked. That being said, as a little excuse for "no mocks yet", the change is battle-tested. :)

**Review by alexmt — APPROVED** [review]
LGTM

**Comment by jannfis** [comment]
Cherry-picked into release-2.3 branch as https://github.com/argoproj/argo-cd/commit/68f17dbe8e5fba398013e092724923f18955b3ad

**Pull request merged**